02 Toolbar/src/editor.tsx

```diff
@@ -1,104 +1,80 @@
 import * as React from 'react';
 declare function require(name:string);
 const MTRC = require('markdown-to-react-components');
+import {textAreaManager} from './common/managers/textAreaManager';
 
 interface Props {
 }
 
 interface State {
-  content : string;
+  editorContent: string;
+  viewerContent: string;
+  shouldUpdateCursor: boolean;
 }
 
 export class EditorComponent extends React.Component<Props, State> {
-
-  textarea : any;
+  textArea : HTMLTextAreaElement;
+  offset: number;
+  cursorPosition: number;
 
   constructor(props: Props) {
     super(props);
 
-    this.state = {content: 'test'};
+    this.state = {
+      editorContent: 'test',
+      viewerContent: 'test',
+      shouldUpdateCursor: false
+    };
   }
 
-  //http://stackoverflow.com/questions/1064089/inserting-a-text-where-cursor-is-using-javascript-jquery
-  //https://developer.mozilla.org/en-US/docs/Web/API/Document/getElementsByClassName
-  //https://facebook.github.io/react/docs/refs-and-the-dom.html
-  // offset, pending IE fix
-  // TODO: Refactor, clean code this
-  // Proposal
-  //
-  // insertAtCaret(txtArea: HMLElement, text: string, offsetCursor = 0) {
-  //  const scrollPos : number = txtarea.scrollTop;
-  //  let strPos : number = 0;
-  //
-  //  strPos = calculateTextAreaCurrentCursorPosition(txtarea);
-  //  insertText(txtArea, strPos, text);
-  //  placeCursor(txtArea, strPos, offset);
-  // }
-  insertAtCaret(txtarea : HTMLElement, text, offsetCursor = 0) {
-  		var scrollPos = txtarea.scrollTop;
-  		var strPos = 0;
-      // TODO: this "ie" flag seems to be for old Internet explorer versions
-      // if it's above 11 it fallbacks into "ff" we can simplify this
-  		var br = ((txtarea['selectionStart'] || txtarea['selectionStart'] == '0') ?
-  			"ff" : (document['selection'] ? "ie" : false ) );
-  		if (br == "ie") {
-  			txtarea.focus();
-  			var range = document['selection'].createRange();
-  			range.moveStart ('character', -txtarea['value'].length);
-  			strPos = range.text.length;
-  		} else if (br == "ff") {
-  			strPos = txtarea['selectionStart'];
-  		}
+  componentDidUpdate() {
+    if (this.state.shouldUpdateCursor) {
+      textAreaManager.placeCursor(this.textArea, this.cursorPosition);
+    }
```

Comment:
Because when react update state and then component, it places cursor to the end of text